src/Npgsql/NpgsqlConnector.cs

```diff
@@ -147,6 +147,11 @@ sealed partial class NpgsqlConnector : IDisposable
         /// </summary>
         volatile Exception? _breakReason;
 
+        /// <summary>
+        /// Semaphore, used to synchronize DatabaseInfo between multiple connections, so it wouldn't be loaded in parallel.
+        /// </summary>
+        static readonly SemaphoreSlim DatabaseInfoSemaphore = new SemaphoreSlim(1);
```

Comment:
One problem with this, is that it prevents parallel type loading to *different* databases. In other words, if I start my program and immediately connect to 10 different databases, there's no reason to do that in serial rather than in parallel. I admit this is probably a rare scenario (and people might not care about the delay), but...

If we want to support the above, we could have a "stub" struct as the value in the concurrent dictionary, which would contain the semaphore and a pointer to the NpgsqlDatabaseInfo. So it would work something like this:

1. Call GetOrAdd on the cache (either get or create the stub)
2. Once you have the stub, if the NpgsqlDatabaseInfo if non-null, return that.
3. If it's null, acquire the semaphore and do type loading

What do you think?

Reply:
> One problem with this, is that it prevents parallel type loading to different databases. In other words, if I start my program and immediately connect to 10 different databases, there's no reason to do that in serial rather than in parallel. I admit this is probably a rare scenario (and people might not care about the delay), but...

Exactly what I've described above. If you do think it's worth supporting, then I guess...

Reply:
In the world of microservices it's not possible, but we saw some multiple tenant application issues. So I would say it's possible, but let's wait for a feedback first and not overcomplicate things.